keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,389 @@
+  // +listType=atomic
+  // +kubebuilder:validation:UniqueItems
+  TargetReplicatedJobs []string `json:"targetReplicatedJobs,omitempty"`
+}
+
+type FailurePolicy struct {
+  // MaxRestarts defines the limit on the number of JobSet restarts.
+  // A restart is achieved by recreating all active child jobs.
+  MaxRestarts int32 `json:"maxRestarts,omitempty"`
+  // List of failure policy rules for this JobSet.
+  // Only one rule can match a unique (Job failure reason, ReplicatedJob)
+  // pair. If such a rule exists, it is executed. Otherwise, the default
+  // failure policy of attempting to restart the JobSet is applied.
+  // +optional
+  Rules []FailurePolicyRule `json:"rules,omitempty"`
+}
+
+```
+
+### Constraints
+
+- Each unique (Job failure reason, target Replicated Job) pair can only be associated with one FailurePolicyRule.
+This gives the user the flexibility to specify different failure and restart semantics for jobs failing for the same
+reason, depending on what Replicated Job they are part of. However, within a Replicated Job, there cannot be more than
+one way of handling a particular Job failure reason.
+
+- OnJobFailureReasons must be a valid Job failure reason as defined [here](https://github.com/kubernetes/kubernetes/blob/2d4100335e4c4ccc28f96fac78153f378212da4c/staging/src/k8s.io/api/batch/v1/types.go#L537-L554)
+in the batchv1 Job API. At the time of writing this KEP, these include:
+  - `PodFailurePolicy`
+  - `BackoffLimitExceeded`
+  - `DeadlineExceeded`
+  - `MaxFailedIndexesExceeded`
+  - `FailedIndexes`
+
+### Implmentation
+
+The core part of the implementation will be defining what specific mechanisms the JobSet controller uses to implement
+the behavior defined for each FailurePolicyAction type:
+
+1) `RestartJob`: To restart a single child job without restarting the entire JobSet, the controller will delete that 
+particular child job and allow the normal reconciliation process to recreate it.
+
+2) `FailJob`: To leave a particular child job in a failed state without restarting it or restarting the JobSet, the
+controller will simply do nothing, taking no action on this job.
+
+3) `FailJobSet`: To fail the JobSet immediately without restarting, the controller updates the JobSet status to failed.
+
+4) `Ignore`: To "ignore" the failure (i.e., restart the JobSet without counting it against `MaxRestarts`), the controller
+will delete all child jobs and allow the normal reconciliation process to recreate them. This is in contrast to how
+we restart the JobSet **without** ignoring the failure, which is done by incrementing the counter on the JobSet annotation
+`jobset.sigs.k8s.io/restart-attempt`, which on subsequent reconiliations will trigger job deletion and recreation for any
+child job with a `jobset.sigs.k8s.io/restart-attempt` counter value less than that of the JobSet.
+
+
+### Test Plan
+
+The testing plan will focus on integration tests. Specific test cases and scenarios are defined in the integration test
```

Comment:
I think you probably should have unit tests tracking this feature also.

You probably want tests for validation and I'm sure there will be some unit tests required for this.